Add missing arrows: self-loops, and "Index" to "Hyper"

eriknw opened the pull request:
@DrTimothyAldenDavis, what do you think?

I think your "Index" followed by "Hyper" (NACHO's "sparse" to "doubly-compressed", or S-DC) makes sense. That is, I disagree with your rule 5.

Rules (1), (3), and (4) could probably be simplified as something like "Full may only follow Index or Full".  You could also try replacing rules 1-4 with something like:
- The final non-"Full" level must be "Index"
- The final "Index" level may be followed by 0 or more "Full" levels

Or try to be even more brief to replace rules 1-4:
- The sparse format must end with "Index" level followed by 0 or more "Full" levels

but perhaps you prefer to exclude each path to "Full" explicitly.

Also, I think rule (6) was supposed to say `(..., Index, Sparse, ...)` instead of `(..., Index, Full, ...)`.

eriknw committed:
Add missing arrows: self-loops, and "Index" to "Hyper"

DrTimothyAldenDavis commented:
My rule 6 has a typo.  That should be `(..., Index, Sparse, ...) `as you point out.

The self-loops are fine.  They are there in the rules but I left them out of the figure so I could draw it more easily.

The arrow from Index -> Hyper is where I still disagree.  This requires adding explicit indices in place of the pointer array.  I think (... Index, Hyper, ...) adds too much complication, and it would be better done as (..., Hyper, Hyper, ...).  That is, add the pointer to change Index into Hyper.

The duplicate indices in the Index -> Hyper transition mean something else than duplicates in other Index categories (where they are duplicates that specify the same entries and can be merged with a dup operator, and they can be out of order).  The duplicates in Index->Hyper can't be removed and must be in order ( I think, ... at least some kind of grouped order).

DrTimothyAldenDavis commented:
Let's talk it over some more.  We're getting close to a workable description, I think.

eriknw commented:
Yeah, we're getting closer :)

`(..., Hyper, Hyper, ...)` has an extra pointers array compared to `(... Index, Hyper, ...)`. Consider the case where the first pointers array with Hyper-Hyper is `[0, 1, 2, ..., N]`. Why would you store that?! It would be more efficient to drop that pointers array and keep the index arrays the same, thereby giving you Index-Hyper. Moreover, why _prevent_ somebody from storing the more efficient version?

Perhaps it would be better to bundle these two indices together and give it a new format name (`HyperCOO(2)`? `MultiHyper(2)`? `TupleHyper(2)`? or just plain `Hyper(2)`?).

Relatedly, a potential advantage of bundling levels together (such as change `Index-Index-Index` to `Index(3)`) is it can let you store this as a 2d array in C or FORTRAN contiguousness.

eriknw committed:
Update rule 6, and undo Index->Hyper

eriknw commented:
I fixed rule 6, and I reverted Index->Hyper. This doesn't mean I'm conceding, I just wanted the diagram to be consistent with the text so we can merge the PR ;)

eriknw commented:
I'll provide an example. Consider this COO data for a rank 4 tensor:
```
# COO-like data
indices = [
    [1, 5, 9],  # axis 0 indices
    [2, 6, 10],  # axis 1 indices
    [3, 7, 11],  # axis 2 indices
    [4, 8, 12],  # axis 3 indices
]
```
For `hyper-hyper-hyper-index`, we have:
```
indices:
    0: [1, 5, 9]
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12]

pointers:
    0: [0, 1, 2, 3]
    1: [0, 1, 2, 3]
    2: [0, 1, 2, 3]
```

For `index-index-hyper-index`, we have:
```
indices:
    0: [1, 5, 9]
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12]

pointers:
    0: NULL
    1: NULL
    2: [0, 1, 2, 3]
```

DrTimothyAldenDavis commented:
No, that's not what (hyper,hyper, hyper, index) would look like for that matrix.

I just posted an example for (hyper,hyper,index).

DrTimothyAldenDavis commented:
I think we might have a mis-match on the matrix format, regardless of how we think about each case.

Either that, or your example is a special case.

eriknw commented:
What is wrong with my example? Yes, it is carefully chosen. Maybe this slightly different example will be clearer:

```
# COO-like data
ndices = [
    [1, 5, 9, 9],  # axis 0 indices
    [2, 6, 10, 10],  # axis 1 indices
    [3, 7, 11, 11],  # axis 2 indices
    [4, 8, 12, 13],  # axis 3 indices
]
```
For `hyper-hyper-hyper-index`, we have:
```
indices:
    0: [1, 5, 9]
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12, 13]

pointers:
    0: [0, 1, 2, 3]
    1: [0, 1, 2, 3]
    2: [0, 1, 2, 4]
```
For `index-index-hyper-index`, we have:
```
indices:
    0: [1, 5, 9]
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12, 13]

pointers:
    0: NULL
    1: NULL
    2: [0, 1, 2, 4]
```

DrTimothyAldenDavis commented:
I think your example is correct -- I misread it.

In general, with (hyper, hyper, hyper, index), the indices and pointers at each axis will differ, and they won't be as simple as [0 1 2 3... n].

DrTimothyAldenDavis commented:
See my example I just posted for the 3D tensor, in (Hyper,Hyper,Index) format (DC-DC-S in your notation, right?).

eriknw commented:
> In general, with (hyper, hyper, hyper, index), the indices and pointers at each axis will differ, and they won't be as simple as [0 1 2 3... n].

In general, yes, but that doesn't mean `index-hyper` combo is invalid or wouldn't be useful. In fact, I think it _is_ useful. It can be convenient to encode information very sparsely in one or more dimensions.

`index-index-hyper-index` can be useful even if there's _some_ duplicated indices. Let me make one change to my previous example:
```
# COO-like data
ndices = [
    [1, 1, 9, 9],  # axis 0 indices
    [2, 6, 10, 10],  # axis 1 indices
    [3, 7, 11, 11],  # axis 2 indices
    [4, 8, 12, 13],  # axis 3 indices
]
```
For `hyper-hyper-hyper-index`, we have:
```
indices:
    0: [1, 9]
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12, 13]

pointers:
    0: [0, 2, 3]
    1: [0, 1, 2, 3]
    2: [0, 1, 2, 4]
```
For `index-index-hyper-index`, we have:
```
indices:
    0: [1, 1, 9]  <-- you don't like or want to allow 1 to be duplicated here, right?
    1: [2, 6, 10]
    2: [3, 7, 11]
    3: [4, 8, 12, 13]

pointers:
    0: NULL
    1: NULL
    2: [0, 1, 2, 4]
```

> See my example I just posted for the 3D tensor, in (Hyper,Hyper,Index) format (DC-DC-S in your notation, right?).

I'll take a closer look to make sure we're referring to the same things. Yes, I would call this DC-DC-S.  I think you know I disagree with this statement:
```
    (Index, Hyper, Index): invalid since the first axis0 has no way
        to refer to variable-sized parts in axis1.
```

DrTimothyAldenDavis commented:
Yes, I think the only thing we now differ on is the `(..., Index, Hyper, ...) ` question.  Well, the names as well, but I'm open to other names than mine (I have Index, Sparse, Hyper, Full names for your S, C, DC, and F (?) names I think).  I don't really mind what we call the formats.

Regarding duplicate indices:

I don't mind if indices are duplicated and/or out of order, it's just that by allowing (..., Index, Hyper, ...) they mean different things in different places.  I would like duplicate indices to always mean there is duplicate content that must be summed, with a dup operator, like how GrB_Matrix_build works.  That's useful, even in a file.  You might want to specify a 3D tensor as a sum of 3-by-3-by-3 dense sub-tensors, for example, which can then overlap at their corners.  Sure, those overlapping entries should eventually be merged, but the specific values may be of interest, before merging, because of the underlying physics, let's say.  So they could be saved as duplicates in a file, and merged later.

With the `(... Index,Hyper, ...)` format, duplicate indices along that particular Index axis mean something else entirely.  They are not representing duplicates, and they must be at least partially ordered (by group I guess).   They are not to be summed with some kind of dup operator; they can't be removed.  That's strange to me, and I think it complicates the format overall.

eriknw commented:
> I would like duplicate indices to always mean there is duplicate content that must be summed

Understood. I think this is what should be relaxed. Allowing duplicate indices that mean "please sum these together" should be an _extension_ to the file format, because I think this complicates things (although I know there are reasons duplicates arise in practice). Vanilla file format should not allow duplicates that need merged imho--having formats that imply computation seem less portable to me.

As an exercise, let's look at level formats in a new way (i.e., new for both of us). Let's try bundling preceding "index" levels (in your parlance) with the other levels. Let me illustrate with examples:
```
HyperSparse(3):
    NACHO: S-S-DC
    TIM: index-index-hyper

Sparse(3):
    NACHO: S-DC-C
    TIM: index-index-sparse

Full(3, 0)
    NACHO: S-S-S
    TIM: index-index-index

Full(1, 2)
    NACHO: S-F-F
    TIM: index-full-full

  /--\
  |  |
  |  v
HyperSparse(n1) --> Full(n3, f) --> Done!
  ^  |               ^
  |  v              /
Sparse(n2) --------/
  ^  |
  |  |
  \--/

n1 > 0
n2 > 0
n3 > 0
f >= 0
```
What do you think? I think the diagram is pretty clean. In this view, your preference of not allowing `index-hyper` is equivalent to requiring `n1=1`.

If desired, we could break e.g. `Full(n, f)` into `COO(n)` and `Full(f)`.

DrTimothyAldenDavis commented:
I'm confused because I thought the right-most axis always had to be S or F (index or full in my naming scheme).

eriknw commented:
We could also break it up as e.g.
```
  /--\              Done!
  |  |               ^
  |  v               |
HyperSparse(n1) --> COO(n3) --> Full(n4) --> Done!
  ^  |               ^
  |  v              /
Sparse(n2) --------/
  ^  |
  |  |
  \--/
```

DrTimothyAldenDavis commented:
still confused.
```

HyperSparse(3):
    NACHO: S-S-DC
    TIM: index-index-hyper
```

isn't that backwards?  Do you mean DC-S-S and hyper-index-index?

eriknw commented:
> isn't that backwards? Do you mean DC-S-S and hyper-index-index?

No. This is bunding _preceding_ "index" levels, such as `index-index-hyper`. With `index-index-hyper`, indices for three levels are stored, and each level has the same number of indices. With proper motivation, one could imagine storing these grouped indices together "rowwise" or "columnwise".

DrTimothyAldenDavis commented:
But index-index-hyper ends in hyper.

eriknw commented:
> But index-index-hyper ends in hyper.

Sorry, I guess I didn't explain things well.  This is like saying "hyper ends in hyper" (duh!). "hyper" isn't a complete sparse structure. It is a single component in a complete sparse structure.  Similarly, `Sparse(n)`, `Hypersparse(n)`, etc. are _components_ of a complete sparse structure. The rules for creating the complete sparse structure are shown in the state diagrams above.

eriknw commented:
Here's a concise, shallow summary of the differences between me and Tim (I think a deeper summary would cover differences of principles). These are the all the pairs of levels that can appear next to each other in a sparse structure:
```
NACHO   Tim             i0 p0 i1 p1
-----------------------------------
C-C     sparse-sparse             X
C-DC    sparse-hyper        X  X  X
C-S     sparse-index        X  X
DC-C    index-sparse(*1) X        X
DC-DC   hyper-hyper      X  X  X  X
DC-S    hyper-index      X  X  X
S-C     N/A(*2)          X        X
S-DC    index-hyper(*3)  X     X  X
S-S     index-index      X     X
S-F     index-full       X
F-F     full-full
```
`i0`: whether indices for the first level in the pair are stored
`p0`: whether pointers for the first level in the pair are stored
`i1`: whether indices for the second level in the pair are stored
`p1`: whether pointers for the second level in the pair are stored

Recall the abbreviations for NACHO: `C=compressed, DC=doubly-comprssed, S=sparse, F=full`.

We agree that sparse structures must end with `index` or `full`, and only `full` can follow `full`.

Note that `C-C` stores a single pointers array that is used for both levels.

We have three differences marked by asterisks above:
- `*1`: we agree on the structure, but our naming conventions differ. All other places, `DC=hyper` and `index=S`, but here we have `DC=index`.
- `*2`: Tim doesn't allow this. `S-C` is, admittedly, a pretty stupid data structure most of the time, because it duplicates data in a regular pattern and would probably be better stored as `DC-C` / `index-sparse`. It's complicated to create and use too. But, who's to say it would never be useful? After all, data is duplicated for the GPU all the time...
- `*3`: I think we should allow this, but Tim has reasons to not like it. This is what we've been discussing the previous several posts.

@DrTimothyAldenDavis, is this a fair and accurate (if incomplete) summary? Does anything need clarified? (btw, thanks for the spirited discussion--I'm glad to have somebody to talk to about this!).

DrTimothyAldenDavis commented:
I'm ok with what I call index-sparse, or what you call S-C. We might disagree on what it means though. In a 10 by 10 by 10 tensor in S-C-S format, the first index_0 array can have size, say, of 3.  Say it contains the list (0,2,5). Then in the tensor there are no entries outside of A([0,2,5],:,:), and pointer_1 has length 31. 

index-hyper doesn't make sense to me. The key reason is more fundamental I think. 

I want to have a simple set of rules that describe how each axis works and connects to the next, without any niche exceptions where one axis changes its data structure when paired with another. I didn't realize for example that your C-C differs from my view of it. 

C for axis k should mean that index_k is never present and pointer_k is always present. 

DC means both always present etc. 

If we have lots of exceptions then the implementation gets very complicated

eriknw commented:
> I'm ok with what I call index-sparse, or what you call S-C. We might disagree on what it means though.

Let's not change what NACHO means in this discussion. That way lies certain madness.  `S-C` means something specific (and perhaps a little silly) in NACHO, and your proposal doesn't support it. Your `index-sparse` is `DC-C` in NACHO. If you look closely, points `*1` and `*2` above are closely related.

> index-hyper doesn't make sense to me.

To clarify, do you understand what I mean by index-hyper? Do you understand my examples? The data structures are well-defined. Or are you caught up that it doesn't fit in well with some of your current principles or goals?

> I didn't realize for example that your C-C differs from my view of it.
> C for axis k should mean that index_k is never present and pointer_k is always present.

For `C-C` / `sparse-sparse`, why would you store the first pointers? That makes no sense whatsoever. It's completely pointless. If stored, the first pointers array would follow a simple pattern, `[0, 10, 20, ...]`, and doesn't need to be used directly.

If we try to keep to "simple" rules, perhaps you would prefer to not allow `sparse-sparse`? Instead, it could be called `full-sparse`, because neither the index nor pointers array needs to be stored for the full dimension here.

I'm particularly interested to hear your thoughts on `C-C` / `sparse-sparse` / `full-sparse`.

> I want to have a simple set of rules [...]. If we have lots of exceptions [...]

I agree with seeking for a simple set of rules with few exceptions. If these rules are (a) not expressive enough, (b) too complicated, or (c) have too many exceptions, then maybe we should try to come up with a different set of rules.

Given our current understanding, it appears that we don't like each others rules. I'm hopeful that we can come up with a _new_ set of rules that we both like.

As a reminder here are the rules for NACHO:
<hr>

We can choose how each sparse dimension is stored with any of the following formats:
- **`S`, "sparse"**: like COO; indices are "aligned" to the following dimension indices or values.
  - Uses: `indices_i`
- **`C`, "compressed"**: like CSR; fast lookup by index into pointers to the next dimension.
  - Uses: `pointers_i`
  - Makes `pointers_{i-1}` unnecessary (you could store them, but why?!)
- **`DC`, "doubly compressed"**: like DCSR; store unique index values and pointers to the next dimension.
  - Uses: `pointers_i`, `indices_i`

Currently, the final sparse dimension must be "sparse", **`S`**, and may be followed by any number of "full" dimensions.

eriknw commented:
I'm going to post a WIP. It probably won't make any sense yet. This continues from the "bundling" examples above.
```
      /--\              Done!
      |  |                ^
      |  v                |
    HyperSparse(n1) --> Coord(n3) --> Full(n4) --> Done!
      ^  |               ^
      |  |              /
      |  v             /
    Sparse(n2, p) ----/
      ^  |
      |  |
      \--/

- n1, n2, n3, n4 are natural numbers, Z+, in the set {1, 2, 3, ...}
  that indicate the number of dimensions in the group.

- HyperSparse stores indices for one or more dimensions (can be a 2d array) and a pointers array.

- Sparse stores indices for zero or more dimensions (can be a 2d array) and a pointers array.
    - p is a natural number indicating the number of dimensions captured by the pointers array.
    - n2 >= p when Sparse is the first dimension group
    - n2 > p when Sparse is not the first dimension group

- Coord stores indices for one or more dimensions (can be a 2d array)

- Full indicates additional "dense" dimensions

- The tensor structure may begin with any group (but if it starts with Full, it won't be sparse)
```
This has some properties we both may like:
- It does not support NACHO's `S-C` (so it gets rid of a complicated edge case that nobody likes)
- It supports all other NACHO formats (including `S-DC` / `index-hyper`, and `DC-C` / `index-sparse`)
- It explicitly handles `C-C` with a single pointers array (via `p` argument)
- Has a simple state machine (imho) for constructing sparse tensors levels
- Allows indices such as for COO to be stored rowwise or columnwise (C or FORTRAN) order

I suspect it may take us both some time to get used to thinking in terms of "grouped" levels (and I probably need to explain more), but I think other people (such as @willow-ahrens) prefer to think in terms of level groups.

eriknw commented:
Other equivalent variations (sorry for not explaining well--these may only be useful during a call):
```
Option 1

      /---\             Done!
      |   |               ^
      |   v               |
    HyperSparse(n1) --> Coord(n4) --> Full(n5) --> Done!
      ^   |               ^
      |   v              /
      | Indices(n2)     /
      |   ^            /
      |   |           /
      |   v          /
    Sparse(n3) -----/

n1, n2, n3, n4, n5 are natural numbers
```
```
Option 2

      /---\
      |   |
      |   v
    HyperSparse(n1)
      ^   |        /--> Done!
      |   |       /      ^
      |   v      /       |
      | Coord(n2) --> Full(n4)
      |   ^
      |   |
      |   v
    Sparse(n3)

n1, n2, n3, n4 are natural numbers

```
```
Option 3

      /---\
      |   |
      |   v
    HyperSparse   /--> Done!
      ^   ^      /       ^
      |   |     /        |
      |   v    /         |
      | Coord(n1) --> Full(n3)
      |   ^
      |   |
      |   v
    Sparse(n2)

n1, n2, n3 are natural numbers
```
Of these three, my gut reaction is to prefer option 2 for the implied data structures and behaviors.

eriknw commented:
I think I like Option 2 from the previous post. It's an interesting middle ground between our approaches.  Here it is in a bit more detail:
```
      /---\
      |   |
      |   v
    HyperSparse(n1)
      ^   |
      |   |       /--> Full(n4)
      |   v      /       |
      | Coord(n2)        |
      |   ^      \       v
      |   |       \--> Done!
      |   v
    Sparse(n3)


n1, n2, n3, and n4 are the number of dimensions of that group.

------
Arrays
------
HyperSparse(n1):
    indices: for n1 dimensions
    pointers

Coord(n2):
    indices: for n2 dimensions

Sparse(n3):
    pointers

Full(n4):
    (no data)

--------------------
Example Combinations
--------------------
(Expressed in NACHO and Tim formats)

HyperSparse(2)-HyperSparse(2)
    S-DC-S-DC
    index-hyper-index-hyper

HyperSparse(2)-Coord(2)
    S-DC-S-S
    index-hyper-index-index

Coord(2)-Sparse(2)
    S-DC-C-C
    index-index-sparse-sparse

Sparse(2)-Coord(2)
    C-C-S-S
    sparse-sparse-index-index

Sparse(2)-HyperSparse(2)
    C-C-S-DC
    sparse-sparse-index-hyper

Coord(2)-Full(2)
    C-C-F-F
    index-index-full-full

-----------
NOT ALLOWED
-----------
Sparse(2)-Sparse(2)
    - Use Sparse(4) instead

Coord(2)-Coord(2)
    - Use Coord(4) instead

Full(2)-Full(2)
    - Use Full(4) instead

Coord(2)-HyperSparse(2)
    - Use HyperSparse(4) instead

HyperSparse(2)-Sparse(2)
    - Use Coord(2)-Sparse(2) instead
```
Observe that no combination permits NACHO's `S-C` (which I think is universally disliked).

eriknw commented:
Let me share one more thing for posterity (I realize this PR probably wasn't the best place for the bigger discussion).

I made a notebook to help visualize the "bundled" groups and comparing to Tim's format and NACHO. Run the notebook here:
https://mybinder.org/v2/gh/GraphBLAS/binsparse-specification/main?filepath=sparsetensorviz/notebooks/Example_Rank4-bundled.ipynb
